### PR TITLE
`emissions_profile_any_compute_profile_ranking()` now handles special cases

### LINKED ISSUE #642: Enhancing profile ranking function in the tiltIndicator package with implementation of profile ranking function in tiltindiactorbefore
Dear Mauro and Kalash,

Could you please ensure that the profile ranking function in the tiltindiactor package works the same as it does in the tiltindicatorBefore package. 

Please note that I think that this change also changes the input data for the tiltIndiacor package as the profile ranking function needs to be run on the whole Ecoinvent dataset 

Could you thereafter also update the toy data set that are needed to run the tiltIndicator package and also could you once all is working remove the profile ranking function in the tiltIndicatorBefore?

Please let me know if something is unclear. 
All the best 
Anne

### PR DESCRIPTION
**Pull request by kalashsinghal**
closes #642 

Dear @maurolepore 
I have shifted `compute_profile_ranking` from tiltIndicatorBefore to tiltIndicator for `emission_profile` and `emission_profile_upstream`. Please use below comments to understand all use cases for ranking the co2 values.

Please note that an R-CMD check is failing in a file which is not in this package. I will work on that check after your review. Thanks!

----

TODO

- [x] [Link related issue/PR]([url](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)). 
- [x] Describe the goal of the PR. Avoid details that are clear in the diff.
- [x] Mark the PR as draft.
- [x] [Include a unit test](https://code-review.tidyverse.org/reviewer/aspects.html#sec-tests).
- [x] Review your own PR in "Files changed".
- [x] Ensure the PR branch is updated.
- [x] Ensure the checks pass.
- [x] Change the status from draft to ready.
- [ ] Polish the PR title and description.
- [x] Assign a reviewer.

EXCEPTIONS

- [ ] Slide here any item that you intentionally choose to not do.

### CODE REVIEW

File: tests/testthat/test-emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [inline tests/testthat/test-emissions_profile_any_compute_profile_ranking.R:138]: Reason to write these these two tests can be understood from this issue: https://github.com/2DegreesInvesting/tiltIndicatorBefore/issues/17

File: tests/testthat/test-emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [inline tests/testthat/test-emissions_profile_any_compute_profile_ranking.R:105]: Reason to write these two tests can be understood from this issue: https://github.com/2DegreesInvesting/tiltIndicatorBefore/issues/18

Comment (kalashsinghal) [inline tests/testthat/test-emissions_profile_any_compute_profile_ranking.R:105]: Reason to write these two tests can be understood from this issue: https://github.com/2DegreesInvesting/tiltIndicatorBefore/issues/18

File: tests/testthat/test-emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [inline tests/testthat/test-emissions_profile_any_compute_profile_ranking.R:154]: Reason to write this test can be understood from this issue: https://github.com/2DegreesInvesting/tiltIndicatorBefore/issues/65

File: R/emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [inline R/emissions_profile_any_compute_profile_ranking.R:58]: I have applied `unique` here to remove extra counts due to duplicate values. For ex: 
Ranks: [3, 2, 2, 1]

Rank proportion without unique: [0.75, 0.5, 0.5, 0.25] (the ranks divided using 4)
Rank proportion with unique: [1, 0.67, 0.67, 0.33] (the ranks divided using 3)

File: R/emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [inline R/emissions_profile_any_compute_profile_ranking.R:97]: There could be a better logic to implement all the above usecases. I haven't much time to improve this logic as it could be the case that you will come up with a better one. Hence, I would like to leave this PR to you :)

**Review by maurolepore — COMMENTED** [review]
Thanks @kalashsinghal, 

I pushed a few refactorings here and some more into [this PR](https://github.com/2DegreesInvesting/tiltIndicator/pull/652), please review it. Once we agree we merge that PR into this one and then  this one into `main`.

File: R/emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [reply R/emissions_profile_any_compute_profile_ranking.R:43]: @maurolepore Thanks for suggesting the above alternatives! I think our `rank_proportion` function best suits our ranking use case.

File: R/emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [inline R/emissions_profile_any_compute_profile_ranking.R]: @maurolepore Please note that I am ranking only the co2 values which are not NA after excluding them using the `filtered_df` dataframe, and then concatenating it in the final output using `bind_rows`. 

And what you are doing with the new code is that you are first ranking all of them and then assigning the NA values to the ones we need. The NA cases should also not be used while ranking the right cases. For example: 

Your current code change: 
isic_4digit: ["12", "4444", "4444", "4444"].   (have to exclude "12" isic)
co2 values: [4, 5, 5, 6]
Rank: [NA, 0.66, 0.66, 1.0]

My previous code: 
isic_4digit: ["12", "4444", "4444", "4444"].   (have to exclude "12" isic)
co2 values: [4, 5, 5, 6]
Rank: [NA, 0.5, 0.5, 1.0]

From the above case you can understand that the 2-digit ISIC should be first excluded for ranking and then needs to be assigned an NA. Please let me know if its clear with this example or not! Thanks!

Comment (maurolepore) [reply R/emissions_profile_any_compute_profile_ranking.R]: Thanks @kalashsinghal, OK, I'll make the change and turn your example here into a test. BTW, this is an R package so I'd appreciate examples in R :-)

File: R/emissions_profile_any_compute_profile_ranking.R
Comment (maurolepore) [inline R/emissions_profile_any_compute_profile_ranking.R:35]: @kalashsinghal 

Here you can see that we first exclude the special cases and only then we compute `profile_ranking`. I also calculate `profile_ranking` for the rows we want to `exclude`  but later I convert them to `NA`. This seems silly but it allows us to extend rather than modify the original code -- following the Open-close (SOLID) principle that "code should be open to extension but closed to modification".

**Review by maurolepore — COMMENTED** [review]
@kalashsinghal,

I made the fix you requested, by which the special cases not only result in assigning `NA` to `profile_ranking` but also they are excluded from the calculation of `profile_ranking`.

Your comment referred only to `*isic_4digit` but I assume you want the same in all special cases. Please see the tests carefully and let me know if that's what you meant or not.

File: tests/testthat/test-emissions_profile_any_compute_profile_ranking.R
Comment (maurolepore) [inline tests/testthat/test-emissions_profile_any_compute_profile_ranking.R:1]: @kalashsinghal please review the tests -- particularly the ones that ensure that `profile_ranking` exclude the special cases you want.

File: R/emissions_profile_any_compute_profile_ranking.R
Comment (maurolepore) [inline R/emissions_profile_any_compute_profile_ranking.R:30]: @kalashsinghal, 

I'm confused. What's the final decision? 

Do we pre-compute `profile_ranking` in tiltIndicatorBefore and pass it to tiltIndicator as a column in the `co2` datasets (i.e. products or inputs)? Correct? 

If so, removing this line will re-compute `profile_ranking` inside the `emissions_profile*()` functions unless we also remove this line:

https://github.com/2DegreesInvesting/tiltIndicator/blob/a220fe46ef6a48406f719a823096a58d09cbdfa0/R/emissions_profile_any_at_product_level.R#L13

File: tests/testthat/test-emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [inline tests/testthat/test-emissions_profile_any_compute_profile_ranking.R:168]: @maurolepore Perfect! Thats what I wanted to happen.

File: R/emissions_profile_any_compute_profile_ranking.R
Comment (kalashsinghal) [reply R/emissions_profile_any_compute_profile_ranking.R:35]: @maurolepore This was a very new logic for me to understand, however its clear now. I agree to the Open-close principle. LGTM

Comment (kalashsinghal) [reply R/emissions_profile_any_compute_profile_ranking.R:35]: And yes I want this change for all the special cases which you did correctly. Please feel free to merge

File: R/emissions_profile_any_compute_profile_ranking.R
Comment (maurolepore) [reply R/emissions_profile_any_compute_profile_ranking.R:30]: @kalashsinghal, 

But then how do we avoid the problem that initiated the whole change? That is, how do we ensure that `profile_ranking` is calculated using a dataset that has products that exist in the entire `ecoinvent` dataset, instead of using a dataset that has products that only exist in europages?

Comment (kalashsinghal) [reply R/emissions_profile_any_compute_profile_ranking.R:30]: @maurolepore There are new datasets namely `emissions_profile_products_ecoinvent` and `emissions_profile_upstream_products_ecoinvent` which has all the ecoinvent products, and they will be ranked in tiltIndictor using the `emissions_profile_any_compute_profile_ranking` function that we have shifted. 

The old datasets namely `pctr_products` and `ictr_products` have products that only exists in `Europages`, and they can also be ranked using the same `emissions_profile_any_compute_profile_ranking` function, however we are not using them for calculating the final results because Anne thinks they needs to be reviewed. 

In either case, we don't need to calculate `profile_ranking` in tiltIndicatorBefore because we have separate datasets for both the cases. Does it make sense?

**Review by maurolepore — APPROVED** [review]

**Comment by AnneSchoenauer** [comment]
Hi both, 
As @kalashsinghal explained this is correct. But I just put it down one more time in my own words: 

- The tiltIndicatorBefore will give datasets that includes ALL ecoinvent products. We then would need to calculate in the tiltIndicator package the profile_ranking for ALL ecoinvent products. 
- However, the risk_categories are then only calculated for those ecoinvent products that were matched with Europages. In other words, we only pick from ALL ecoinvent produts the products that also exist in Europages. That means that the profile_ranking column still refers to ALL ecoinvent products still BUT we only care about the Eruopages products. 
- This is on product level. For the company level nothing will change. 
- @kalashsinghal  for your latest question on slack. In the Output files we also need the profile_ranking column. 

I hope this makes sense? 
If not let me know and I try to clarify it again.  
Best
Anne

**Pull request merged**
